### ISSUE TITLE
Integrate Error Chaining in a js-errors or @matrixai/errors package

### ISSUE THREAD

**Issue by CMCDragonkai** [MatrixAI/Polykey#304]
### Specification

Now that `@matrixai/errors` is ready, we can use it to incorporate error chaining into PK. This will involve a series of steps:

#### Chaining

In all places where we are catching an error and then throwing a new error in its place, we need to be including the original error as the `cause` of the new error, e.g.
```ts
} catch (e) {
  throw new SomePolykeyError(e.message, { cause: e });
}
```

Note that the `cause` should always be an exception/error. If not specified it defaults to `unknown`. When a error is thrown, the top-level error should contain the full instances of the errors in its cause chain, as one big nested object. This data structure can be serialised and deserialised recursively, where every error in the chain has a single cause, and that cause is contained as a property within it.

It is important to realise that the `replacer` will remove any entry if the returned value is `undefined`. Unless it is an array, it will replace that value with `null`. This means that because our `cause` may be `undefined`, it may not exist during deserialisation. Any usage of `fromJSON` must be aware that the `cause` property may not exist.

#### Client Server Error Architecture

```
toError: ServiceError -> ErrorPolykeyRemote                         fromError: Error -> ServerStatusResponse

                                                                         ┌─────────────────────────────┐
      ┌───────────────────────────────┐                                  │ NodeId 1 - Server           │
      │ Client                        │             Request              │                             │
      │                               │                                  │ ┌─────────────────────────┐ │
      │ Receives toError ServiceError ├──────────────────────────────────┼─► Service Handlers        │ │
      │ Throws ErrorPolykeyRemote     │                                  │ │                         │ │
      │ cause: ErrorX | ErrorUnknown  ◄──────────────────────────────────┼─┤ Sends fromError ErrorX  │ │
      │ data: { nodeID: 1, ... }      │                                  │ └─────────────────────────┘ │
      └───────────────────────────────┘      Response with error         │                             │
                                             JSON string in metadata     └─────────────────────────────┘
                                          error: { type: ErrorX, ... }
```

It's important to realise:

* Only the client constructs `ErrorPolykeyRemote`
* The `ErrorPolykeyRemote` should contain `nodeId: NodeId`, `host`, `port` and other connection information in its own `data` which can aid debugging which server responded with this error
* The `ErrorPolykeyRemote` should also have information about which call triggered this, and perhaps even request information, it can have a lot of information which can add later.
  - Additional client/connection metadata should be set in the `data` property of `ErrorPolykeyRemote`
  - Give `data` a more specific type like `data: ClientMetadata & POJO` so that it forces that `ClientMetadata` must be available
  - We can use `type ClientMetadata = { nodeId: NodeId; host: Host; port: Port }`. And additional information can be provided later. Like what the call was.
  - The `ClientMetadata` type should be in `src/types.ts` to avoid import loops, since it has to use the `NodeId` type from `nodes` and `Host` and `Port` from the the `network` domains.
  - This keeps the type along with `ErrorPolykeyRemote` together at the top level `src`.
* The server only serialises the `ErrorX` that `ErrorX` may be any kind of error. The server must filter the information it sends to the client to ensure it is not leaking sensitive information. This can include the `stack` since the `stack` is rarely required by the client.
* Even when the server sends the error back to the client, it must also log out this error, but only if the error is considered a "server error". That is if the is a "client error" such as 4xx errors in HTTP, then it should not be logged out. This includes:
  - Authentication
  - Validation
  - Precondition
  - Concurrency Conflict
  - Anything else coming from HTTP 4xx
  - Example: missing object/resource
* If the client encounters unknown data during deserialisation, it only returns `ErrorUnknown` at the root of the JSON data. Any other unknown data should be returned as-is.
* This means all remote errors are **always** `ErrorPolykeyRemote`, but the `cause` may be `ErrorUnknown` or `ErrorX`. Where `ErrorX` may contain anything after that (as long as the runtime schema checking during `fromJSON` calls work)
* The GRPC metadata key for the serialised error JSON should just be `error`, in the future if we switch to using JSON RPC, this error JSON will probably be encoded as part of the JSON RPC protocol - see https://www.jsonrpc.org/specification#error_object and https://eth.wiki/json-rpc/json-rpc-error-codes-improvement-proposal
* GRPC itself only supports "HTTP2" kind of errors, but these codes aren't really being used, all errors from the remote node is considered one kind of HTTP error, that being of `UNKNOWN`. We use that to represent an "application" error. This basically means we don't use most of GRPC/HTTP2's standard error codes, as our "application layer errors" are on top. This is because GRPC/HTTP2's standard error codes were designed more for the HTTP layer. Although some of the HTTP error codes are still used internally by GRPC.
* This is the same philosophy taken by JSON RPC over HTTP: http://www.simple-is-better.org/json-rpc/transport_http.html, where success or error at the application layer is still 200 response code at the HTTP layer. HTTP errors are specific to the HTTP layer.
  ```
  200 OK
  for responses (both for Response and Error objects)
  204 No Response / 202 Accepted
  for empty responses, i.e. as response to a Notification
  307 Temporary Redirect / 308 Permanent Redirect
  for HTTP-redirections (note that the request may not be automatically retransmitted)
  405 Method Not Allowed
  if HTTP GET is not supported: for all HTTP GET requests
  if HTTP GET is supported: if the client tries to call a non-safe/non-indempotent method via HTTP GET
  415 Unsupported Media Type
  if the Content-Type is not application/json
  others
  for HTTP-errors or HTTP-features outside of the scope of JSON-RPC
  ```

#### JSON serialisation/deserialisation

Our gRPC `toError` and `fromError` utilities will need to be modified to be able to serialise and deserialise the error chain when sending errors between clients/agents or agents/agents.

- These can utilise JSON's replacer/reviver properties of stringify/parse - this is non-recursive and processes objects in fragments, with everything being orchestrated by stringify/parse
- The replacer needs to serialise errors into a standardised structure `{ type: 'ClassName', data: { ... } }` and also filter out sensitive data (the stack) when errors are being sent to an agent (as opposed to a client) - it works top to bottom, creating a structure first and filtering out unwanted fields afterwards
- The reviver needs to covert this structure back into its original data type - it works bottom to top, converting fields to their correct types and then finally constructing the finished data type
  - If the received data can't be deserialised into a known error type it should be converted to an `ErrorPolykeyUnknown`, with the unknown data in the `data` field
- We also no longer need to separate errors in the gRPC metadata into `name`/`message`/`data` and this is now just one data structure - this can just be one `error` field
- Need to filter out sensitive information when sent from the service handlers, this includes `stack` and other special data. This would a case by case basis for specific exception classes. That is, the `replacer` will need to execute a filter by checking against specific exception classes and filtering them out. Filter rules should be acquired from `src/client` and `src/agent`.

#### Logging of Errors at the Service Handlers

- Loggers should be added to the service handlers in order to log out errors, since these do not shut down the agent process
  - Not all errors should be reported though - only validation/authentication/other precondition errors
  - This logger is separate from the logger used for grpc internals
  - How to determine if an error is for the client or for the server? This is a case by case basis, unlike HTTP, we do not have 4xx and 5xx codes emitted at the beginning, all exceptions assume programmatic usage. Therefore there is a smaller amount of client errors then there are server errors. The filter should just check that if part of "client errors" set, then don't emit. Create set of "client errors", this set can be defined in both `src/client` and `src/server` (code duplication is fine here), since each service can define their own set of what is considered a "client error".

#### Reporting the Error to the user at the Root Exception Handler

We have 2 root exception handlers, one at the client and one at the agent. This addresses the client side, however unless things are different, the same applies to the agent side.

On the client side, this is done in `src/bin/polykey.ts`.

When an exception is received, we must interpret 3 things:

* The `cause` chain
* The `exitCode`
* The metadata in `data`

According to issue #323, the `--format json` doesn't currently affect the STDERR. This needs to be done now because the errors being reported are quite complex, and during testing, we expectation utilities should be parsing JSON to make it easier to test.

Therefore `binUtils.outputFormatter` will need to ensure that `error` type is a human formatted JSON, while `json` type can now be the JSON output for exceptions. This is doable now due to the `toJSON` utility function inherited from `AbstractError`.

Furthermore in order to acquire the options passed into the command, use `rootCommand.opts()`, this method was added into to `PolykeyCommand` to enable us to acquire the options.

```ts
const opts = rootCommand.opts();
if (opts.format === 'json') {
  process.stderr.write(
    binUtils.outputFormatter({
      type: 'json',
      data: e.toJSON(),
    });
  );
} else {
  // use `error` format
  process.stderr.write(
    binUtils.outputFormatter({
      type: 'error',
      // ...
    });
  );
}
```

The desired format for human format should be something like:

```
ErrorName: description - message
  K\tV
  K\tV
  cause ErrorName: description - message
    K\tV
    K\tV
```

Note the usage of `\t` for separation while spaces are used for indentation. We can change this format later.

The `exitCode` is currently a bit ambiguous, it was originally intended to mean that if this exception was the last exception caught by the process, this code should be used for the process exit code.

It seems more ideal to allow the process to decide what the exit code should be based on the family of exceptions. But since we have built our exit code this way, we need to continue to use it like this.

However now that we have a `cause` chain, we have to decide how to deal with the exit code when we are getting `ErrorPolykeyRemote`. Because of this scenario:

```
C -> A1 -> A2
```

We must differentiate exceptions that originate from the client, or the first agent or the second agent. If an exception comes from A2, we may see `ErrorPolykeyRemote` wrapping another `ErrorPolykeyRemote`.

So right now a policy can be made:

1. Find the first `cause` property that is not `ErrorPolykeyRemote`, and use that `exitCode`
2. Otherwise get the `exitCode` of the first exception.

This only works because `ErrorPolykeyRemote`'s cause type is limited to `ErrorPolykey`, it cannot be undefined or anything else.

### Additional context

* https://github.com/MatrixAI/js-db/commit/5172d01e694e7f0f58023fb41c76844da0a6284f
* https://github.com/mathew-kurian/TraceError.js - can use this stack style for reporting
* https://github.com/MatrixAI/js-logger/issues/3 - Needs design integration with js-logger so these 2 libraries would work together
* https://github.com/MatrixAI/js-polykey/pull/321#issuecomment-1028563774 - discussion on error handling between local node and remote node for agent & client service, realising the need for additional metadata indicating where the exception comes from, as well as handling non-ErrorPolykey errors
* https://github.com/MatrixAI/js-polykey/pull/311#issuecomment-1049484685 - discussion of how error chaining will make debugging of exceptions that originate from outside the domain being tested easier
* https://github.com/MatrixAI/js-errors/pull/1#issuecomment-1086832650 - Initial ideas on error serialisation and deserialisation in this thread

### Tasks

- [x] 1. Convert all errors from `CustomError` to new `AbstractError<T>` from `@matrixai/errors`
- [x] 2. Give all errors a static description and `errCode`
- [x] 3. In all places where one error is caught and a different one is thrown, include the original error as the `cause`
- [x] 4. Modify the validation utils to use new error chaining
- [x] 5. Modify gRPC `fromError` and `toError` functions to properly handle chained errors
- [x] 6. Include a timestamp when any error is thrown
- [x] 7. Add error logging to service handlers
- [x] 8. Update all packages relying on `js-errors` to `1.1.0` to use the `fromJSON` and `toJSON`
- [x] 9. PK should adapt its replacer and reviver functionality and bring in "rules" from both `src/client` and `src/agent`. Can use DI in the client service vs the agent service to change the serialisation/deserialisation.
- [x] 10. Logging should also use separate rules, one for client service, one for agent service.
- [x] 11. Handle the error chain at the root exception handler for `src/bin/polykey.ts` and `src/bin/polykey-agent.ts`
- [x] 12. Ensure that `AbstractError.fromJSON` can handle non-existent `cause` property

**Comment by CMCDragonkai** [comment]
In our root exception handler, or our `ExitHandlers`, when we finally log out the error, we may wish to use JSON stringification of the exception to provide full context. Note that we don't use `logger.error`, it goes straight to STDERR.

**Comment by CMCDragonkai** [comment]
Error chaining would be useful in a range of areas. Cause right now we are losing error context. One example of this is in the networking when an error event is emitted to the TLS socket, this gets printed out, but it's not clear what this actual error is and where it is truly coming from.

**Comment by CMCDragonkai** [comment]
Here's an example problem from: https://github.com/MatrixAI/js-polykey/pull/321#issuecomment-1028554814

In our GRPC `fromError`, we are sending over serialised `ErrorPolykey` instances.

However not all errors are going to be `ErrorPolykey`. Therefore to deal with those errors, we should wrap them in `ErrorPolykey`. It would be ideal to be able to construct an error chain here, so that it is possible to tell the client:

1. An error occurred on a remote node during a call
2. Here is the error that actually occurred on the remote node.

In fact, this could be done for `ErrorPolykey` instances and child instances as well in order to also capture network/node metadata so it's obvious that remote errors are different from local errors. 

Meaning the resulting exception being thrown might actually be a specific exception like `ErrorPolykeyRemote`. This can allow us to wrap additional data here like `nodeId` and handler name/location, maybe even Polykey version.

And it's the chain itself dispatched in `toError` to all the possible error instances.

**Comment by CMCDragonkai** [comment]
One issue is the serialisation of the cause chain.

The metadata doesn't allow any kind of nested structure. However all the metadata properties can be multi-values.

This means that one could flatten the chain so that each error property goes into the array to represent the error chain. 

```ts
for (const e of errorChain) {
  metadata.add('name', e.name);
  metadata.add('message', e.message);
  // This needs to be considered as well
  if (e.data != null) {
    metadata.add('data', JSON.stringify(e.data));
  }
}
```

**Comment by CMCDragonkai** [comment]
Remember that `grpc` domain is our "usage" of GRPC, not a general GRPC library, so we can always make changes there to suit our usecase of GRPC.

**Comment by emmacasolin** [comment]
> One issue is the serialisation of the cause chain.
> 
> The metadata doesn't allow any kind of nested structure. However all the metadata properties can be multi-values.
> 
> This means that one could flatten the chain so that each error property goes into the array to represent the error chain.
> 
> ```ts
> for (const e of errorChain) {
>   metadata.add('name', e.name);
>   metadata.add('message', e.message);
>   // This needs to be considered as well
>   if (e.data != null) {
>     metadata.add('data', JSON.stringify(e.data));
>   }
> }
> ```

I think this could work, and then in `toError()` instead of doing
```ts
const errorName = e.metadata.get('name')[0] as string;
const errorMessage = e.metadata.get('message')[0] as string;
const errorData = e.metadata.get('data')[0] as string;
```
We could just iterate through them all, returning an array of errors rather than a single one. I'm not sure if this would maintain the correct order though. It would then be up to https://github.com/MatrixAI/js-polykey/issues/323 to decide how this array of errors is displayed.

**Comment by CMCDragonkai** [comment]
Yea that should maintain the chain order because the multivalues are iterated in order of insertion.

**Comment by CMCDragonkai** [comment]
Slightly different from the `ValidationError` problem though cause this required a special way of encoding.

Another way is to use JSON encode for all error properties rather than spreading it into the GRPC metadata. This way we could use `toJSON` magic method to specialise serialised encoding for specific exceptions where applicable.

**Comment by CMCDragonkai** [comment]
Will be done here: https://github.com/MatrixAI/js-errors

**Comment by CMCDragonkai** [comment]
From this: https://github.com/MatrixAI/js-polykey/pull/357#issuecomment-1085463939

It looks like it would be useful debugging to have a timestamp set for when an exception is constructed. Ideally also when it is thrown, but there's no hook for this in the JS runtime I can find.

This is because there's the time when an exception is constructed, an arbitrary delay, when it was thrown, an arbitrary delay, and finally when an error is actually reported using the js-logger.

We would need high precision for this, and not just unixtime. We can use `performance.now()` to help with this rather than just `Date.now()`.

**Comment by CMCDragonkai** [comment]
Being resolved in https://github.com/MatrixAI/js-errors/pull/1

**Comment by CMCDragonkai** [comment]
Not resolved until `js-errors` is actually integrated into PK.

**Comment by CMCDragonkai** [comment]
Integrated into `@matrixai/db` now as `3.2.0`.

Still to do:

1. [x] `@matrixai/async-locks` - https://github.com/MatrixAI/js-async-locks/pull/4
2. [x] `@matrixai/async-init` - should be done along with async-locks - https://github.com/MatrixAI/js-async-init/pull/8
3. [x] `@matrixai/workers` - should be done along with async-init - https://github.com/MatrixAI/js-workers/pull/2
4. [x] `encryptedfs` - should be done along with `js-resources` integration - https://github.com/MatrixAI/js-encryptedfs/pull/63

**Comment by CMCDragonkai** [comment]
Any generic serialisation and deserialisation of our error chain will be done here in PK, it was too difficult to figure out how to do it generically in the `@matrixai/errors` package.

**Comment by CMCDragonkai** [comment]
The DB upgrade will be a fairly large upgrade, a good test of the entire DB transaction overhaul too #366.

**Comment by CMCDragonkai** [comment]
Any serialisation and deserialisation of the error chain should be using the `reviver` and `replacer` pattern. That's where some data may be stripped. That is from agent to agent, some exception data may be sensitive, and so that information should be stripped. However it's better that such information never gets reported in the first place.

Examples include things like password or tokens being incorrect, never report the plaintext password in the exception.

Any information that is considered "secret" should not be put into the exception properties. These include:

* Tokens and passwords (session or temporary tokens)
* Vault secrets
* Network information (unless public addresses/ports)
* Metadata that may enable cryptographic breakthrough

If an exception is going to refer to such a property, they should just refer via a "reference". Like "the password", rather than "password abc123".

To be very safe, we can just eliminate all internal error information when passing error from agent to agent. That is we "collapse" such information into a set of general error codes so that the details of the exception are only privy to the user of the PK agent, not other PK agents. These errors are ultimately logged out on STDERR anyway.

**Comment by CMCDragonkai** [comment]
I think we should both be wary of putting sensitive information into exception properties and logs, AND also clear the "exception chain" and other properties when going from agent to agent.

In that case, there would be important exceptions that communicate legitimate reasons for an error:

1. Something is missing
2. Some precondition is not fulfilled
3. Some system error occurred
4. Some validation error

These sorts of errors we do want the sending agent to receive. We could create list of allowed exception classes to preserve.

This is where the HTTP codes provide a good guideline.

I can imagine a case where:

```
Agent1 (acting as client) <-> Agent2 (acting as server)
                              throw new SpecialError()
                              wrap as NetworkError(specialError)
                              respond with NetworkError
receive NetworkError
unwrap NetworkError(specialError)
handle specialError
```

The `NetworkError` is therefore an intermediate wrapper only intended to carry additional information about the remote node. The handling of the error falls upon the client's request, to handle the fact that an error came from the other agent, and that the actual error is the special error.

See also: https://rebolon.medium.com/json-reviver-pattern-how-to-restore-object-from-json-into-real-entity-737b1a220450

**Comment by emmacasolin** [comment]
So is the `replacer` converting a `PolykeyError`/other error into a `NetworkError` in `fromError` on the sending side and the `reviver` converting a `NetworkError` back into a `PolykeyError`/other error in `toError` on the receiving side? Also I'm a little confused about whether the exception chain should be cleared or not during this process.

**Comment by emmacasolin** [comment]
Previously our `ErrorParse` validation error was using `CustomError`, however, if we want to switch this over to use `AbstractError<T>` then we need to provide a type argument (for the cause).

This is the old version of `ErrorParse`:
```ts
class ErrorParse extends CustomError {
  public keyPath: Array<string>;
  public value: any;
  public context: object;
  constructor(message?: string) {
    super(message);
  }
}
```

It stores a `keypath`, `value`, and `context` as public properties. Since `AbstractError<T>` has inbuilt options for `data` and `cause` properties, we could set the `keypath` and `context` inside the `data` property and `value` as the `cause` (as an example)?

Similarly with `ErrorValidation`, we previously needed to use a bit of a patch fix to make it compatible with our GRPC error serialisation:
```ts
class ErrorValidation extends ErrorPolykey {
  description = 'Input data failed validation';
  exitCode = sysexits.DATAERR;
  public errors: Array<ErrorParse>;
  constructor(message, data) {
    super(message, data);
    if (data.errors != null) {
      const errors: Array<ErrorParse> = [];
      for (const eData of data.errors) {
        const errorParse = new ErrorParse(eData.message);
        errorParse.keyPath = eData.keyPath;
        errorParse.value = eData.value;
        errorParse.context = eData.context;
        errors.push(errorParse);
      }
      this.errors = errors;
    }
  }

  /**
   * This packages an `ErrorParse` array into the `data` property
   * This is to allow encoding to and decoding from GRPC errors
   */
  static createFromErrors(errors: Array<ErrorParse>): ErrorValidation {
    const message = errors.map((e) => e.message).join('; ');
    const data = {
      errors: errors.map((e) => ({
        message: e.message,
        keyPath: e.keyPath,
        value: e.value.valueOf(),
        context: e.context,
      })),
    };
    const e = new ErrorValidation(message, data);
    e.errors = errors;
    return e;
  }
}
```

Rather than needing a separate `errors` property and a `createFromErrors` method, we could instead now utilise the `data` and `cause` properties of `AbstractError<T>`, perhaps having the `cause` simply be an array of `ErrorParse`s?

**Comment by CMCDragonkai** [comment]
I propose:

```
class ErrorParse extends AbstractError<void> { ... }
```

This is because `ErrorParse` is sort of a self-contained error. It arises from a parse failure of some sort. But the parsing failure is explicit. It is the purpose of the function to figure out whether it parses correctly or not. Therefore it's fine to tell `AbstractError` that the cause will be `void`.

This idea is derived from the principles in Python: https://gist.github.com/CMCDragonkai/08266b1463158f4156f66d4bf077add6. It would be called an "EXCEPTION OVERRIDE".

As for `ErrorValidation`, it should extend the normal polykey hierarchy, and it shouldn't use an array of errors. The `cause` should always (with some exceptions) be another exception.

As for serialisation/deserialisation, you'll need to create a custom reviver and replacer for JSON to handle these exception classes. You would need strict expectations that the exception being received has to be a certain class, or certain type. If not, it would need to be handled by defaulting it to an "unknown case": https://github.com/MatrixAI/js-polykey/issues/304#issuecomment-1109445786 and see: https://github.com/MatrixAI/js-errors/pull/1#issuecomment-1087016212

**Comment by CMCDragonkai** [comment]
We have some flexibility btw:

```
// propagate the "cause", this allows ErrorX to be generic over the cause, and the cause is determined by the instantiator
class ErrorX<T> extends AbstractError<T> { }

// specify the cause type directly, such that ErrorX always has the same cause type (usually used where Y is void)
class ErrorX extends AbstractError<Y> { }

// default the cause type, but allow the instantiator to override it
class ErrorX<T = Y> extends AbstractError<Y> { }
```

Remember `Y` could also be other types, but you should choose between something inheriting `Error`, or `void` or `unknown` or a union of them.

**Comment by emmacasolin** [comment]
> As for `ErrorValidation`, it should extend the normal polykey hierarchy, and it shouldn't use an array of errors. The `cause` should always (with some exceptions) be another exception.

By this do you mean to just keep it how it is currently, with the array of errors being stored in an `errors` property rather than changing to putting it as the cause? Since the whole purpose of `ErrorValidation` is to be a wrapper for an array of `ErrorParse`s so it needs to store them somewhere

**Comment by CMCDragonkai** [comment]
Yes, keep it as `errors` property. The cause is supposed to allow you to traverse an "error chain". Putting an array there stuffs up the whole system.

**Comment by CMCDragonkai** [comment]
Don't forget to make all of your descriptions `static` properties now too. I had changed that in `js-errors`.

**Comment by CMCDragonkai** [comment]
> So is the `replacer` converting a `PolykeyError`/other error into a `NetworkError` in `fromError` on the sending side and the `reviver` converting a `NetworkError` back into a `PolykeyError`/other error in `toError` on the receiving side? Also I'm a little confused about whether the exception chain should be cleared or not during this process.

I think I called it `ErrorPolykeyRemote`. Not entirely sure if this is only agent to agent, or covers client to agent.

The `replacer` is on the `JSON.stringify` which is converting an exception into JSON. That's on the error emitting side (server side).

The `reviver` is on the `JSON.parse` which is converting JSON back into an exception. That's on the error receiving side (client side).

The `grpcUtils.fromError` is on the server side (grpc from error).

The `grpcUtils.toError` is on the client side (grpc to error).

---

You have to first spec out the `ErrorPolykeyRemote` as per https://github.com/MatrixAI/js-errors/pull/1#issuecomment-1087016212, and then work out the proper reviver and replacer code.

It should be strict such that any deviation should result in an "unknown error". Consider you're going over an untrusted network.

**Comment by CMCDragonkai** [comment]
For client to agent, error chain should be preserved. For agent to agent, it depends (we have to work that out separately).

When I was working on this problem I was stuck on how to deal with exception classes that don't exist on the client side. The only way to deal with this is with a base case.

The base case can be either a "unknown error" class, or void.

**Comment by emmacasolin** [comment]
> We have some flexibility btw:
> 
> ```
> // propagate the "cause", this allows ErrorX to be generic over the cause, and the cause is determined by the instantiator
> class ErrorX<T> extends AbstractError<T> { }
> 
> // specify the cause type directly, such that ErrorX always has the same cause type (usually used where Y is void)
> class ErrorX extends AbstractError<Y> { }
> 
> // default the cause type, but allow the instantiator to override it
> class ErrorX<T = Y> extends AbstractError<Y> { }
> ```
> 
> Remember `Y` could also be other types, but you should choose between something inheriting `Error`, or `void` or `unknown` or a union of them.

I've actually just noticed a couple of places where `ErrorParse` can be thrown after another error is caught, for example in `parseSeedNodes` a `TypeError` or `ErrorParse` can be caught and a new `ErrorParse` thrown. For these cases would we want to be able to have these original errors as the cause (thus requiring us to use `class ErrorX<T = Y> extends AbstractError<Y> { }`)? Would this interfere with being able to have an array of `ErrorParse`s if not all of them have `void` as the type argument?

**Comment by CMCDragonkai** [comment]
Yea in those cases, I follow the "EXCEPTION OVERRIDE" pattern. In that sense I'm dropping what errors came from before, and overriding as its own higher-level parse error.

https://gist.github.com/CMCDragonkai/08266b1463158f4156f66d4bf077add6

**Comment by CMCDragonkai** [comment]
@emmacasolin can you work out a spec for the error serialisation/deserialisation as well and update the issue description accordingly.

**Comment by CMCDragonkai** [comment]
The filtering of sensitive details can be done in the `grpcUtils.toError`, and this has to be parameterised by whether it is being called by agent service handlers or client service handlers. The agent service handlers can pass a boolean parameter to indicate it should be filtered.

**Comment by emmacasolin** [comment]
I've noticed that a lot of the errors thrown by the network utils include information in their `data` property that could be considered sensitive? For example
```ts
throw new networkErrors.ErrorCertChainKeyInvalid(
  'Chain certificate public key does not generate its node id',
  {
    data: {
      cert,
      certIndex,
      nodeId: certNodeId,
      commonName: commonName.value,
    },
  },
);
```

Should this information be cleared by the replacer? Or should it not be included in the first place?

**Comment by CMCDragonkai** [comment]
The `nodeId` is public information, so is `commonName`. As for `cert` and `certIndex`, I believe these are all "public information" as certificates are public. I'm not sure about `certIndex`, what is that?

Generally most `data` should be cleared, if they aren't going to be useful to the client side. But even if we "clear it" when sent back to the client side, the server side should still log out the unoriginal unfiltered exception.

So we should be logging out all exceptions that go through `fromError`. That way we know when an agent has had to answer back with an error.

**Comment by emmacasolin** [comment]
I think `certIndex` is the index of the certificate being verified in the certchain. So it's probably information that would be useful for debugging so long as it's safe to potentially send to another agent.

**Comment by CMCDragonkai** [comment]
> So we should be logging out all exceptions that go through fromError. That way we know when an agent has had to answer back with an error.

Actually upon second thought, no don't do this cause this pollutes the behaviour of `fromError` too much. Instead this should be done outside, in a separate function. This can be addressed later. See: #323

**Comment by emmacasolin** [comment]
My previous thinking was that the stripping of sensitive data only needed to occur on the sending `fromError` side (`JSON.stringify`), however, this assumes that all errors thrown by the agent service go through `fromError`, which is not the case for read streams. An error thrown by a readable stream only goes through `toError`, which uses `JSON.parse`, so stripping sensitive data will need to be able to occur in both `fromError` and `toError`, i.e. in both the replacer and the reviver.

**Comment by CMCDragonkai** [comment]
Read streams come from the client, there's no need to strip exception data from the read stream.

Do you mean write streams? Because that also uses `fromError` too.

Remember that `generatorWritable` is what calls `fromError`. That is also used in `generatorDuplex`. Errors can only go from server to client, not the other way around.

So only `fromError` needs to be done. No need to change anything about `toError` in regards to filtering rules.

**Comment by CMCDragonkai** [comment]
Some points to clarify:

The `ErrorPolykey.toJSON` method is currently incorrect. The `toJSON` method of any class should actually be returning a POJO, not a string. So the correct returned structure should look like this:

```
class X {
 public toJSON(): { type: string; data: any } {
  return {
    type: this.name,
    data: { ... },
  };
}
```

This was mentioned here: https://github.com/MatrixAI/js-errors/pull/1#issuecomment-1086832650

Now for this new exception `ErrorPolykeyRemote`. We can put this in `src/errors.ts` as a top-level exception.

```
class ErrorPolykeyRemote {
  // Imagine this gets called in grpcUtils.toError
  // this takes POJO
  public static fromJSON (json: object) {
    // convert JSON to `ErrorPolykeyRemote`
  }

  // Imagine this gets called in grpcUtils.fromError
  // this returns POJO
  public toJSON(): object {
    // convert `ErrorPolykeyRemote` to JSON
  }
}
```

This means `ErrorPolykeyRemote.fromJSON(errorPolykeyRemote.toJSON())` should work (type-check). They are duals of each other.

In `agent/utils.ts`, it should contain filtering rules. Filtering rules can be done on a "class by class basis".

```
// filtering rules is class name to lambda, lambda being the filtering function that is applied to the data
const filteringRules = {
  [Class.name]: (data) => data 
};
```

Applying the filtering rules should be an optional parameter to `grpcUtils.fromError`. Since it used in 2 places, responding to unary calls, or responding in streams, then the parameter must be propagated upwards to `generatorWritable` and `generatorDuplex`.

The reviver and replacer are options sent to `JSON.stringify` and `JSON.parse`. These 2 functions are not called in the `ErrorPolykeyRemote`, but instead called in `grpcUtils.fromError` and `grpcUtils.toError`. These 2 rely on the `errors.ErrorPolykeyRemote` too. Therefore the `reviver` and `replacer` should be defined either in `src/errors.ts` so that these are coupled together with error definitions, or in `src/grpc/utils.ts` with the `fromError` and `toError` functions. Not sure right now, but the former looks more correct to me.

**Comment by CMCDragonkai** [comment]
@emmacasolin please remember to update the spec of this issue as well to be comprehensive:

1. Important to specify the cause chain semantics, and how it's supposed to be an exception. What happens when the cause type is unrecognised?
2. The usage of replacer and reviver represents non-recursive fragment processing functions of the overall tree structure. The knot is tied when you use it with `JSON.parse` and `JSON.stringify`.
3. What is the "base exception" to represent unknown error?
4. The JSON format should be specced out as `{ type: 'ClassName', data: { ... } }`.
5. The usage of `error` in the grpc metadata  property instead of `data`.
6. All service handling errors should also be logged out since they don't shutdown the agent process.

**Comment by emmacasolin** [comment]
> Some points to clarify:
> 
> The `ErrorPolykey.toJSON` method is currently incorrect. The `toJSON` method of any class should actually be returning a POJO, not a string. So the correct returned structure should look like this:
> 
> ```
> class X {
>  public toJSON(): { type: string; data: any } {
>   return {
>     type: this.name,
>     data: { ... },
>   };
> }
> ```
> 
> This was mentioned here: [MatrixAI/js-errors#1 (comment)](https://github.com/MatrixAI/js-errors/pull/1#issuecomment-1086832650)
> 
> Now for this new exception `ErrorPolykeyRemote`. We can put this in `src/errors.ts` as a top-level exception.
> 
> ```
> class ErrorPolykeyRemote {
>   // Imagine this gets called in grpcUtils.toError
>   // this takes POJO
>   public static fromJSON (json: object) {
>     // convert JSON to `ErrorPolykeyRemote`
>   }
> 
>   // Imagine this gets called in grpcUtils.fromError
>   // this returns POJO
>   public toJSON(): object {
>     // convert `ErrorPolykeyRemote` to JSON
>   }
> }
> ```
> 
> This means `ErrorPolykeyRemote.fromJSON(errorPolykeyRemote.toJSON())` should work (type-check). They are duals of each other.
> 
> In `agent/utils.ts`, it should contain filtering rules. Filtering rules can be done on a "class by class basis".
> 
> ```
> // filtering rules is class name to lambda, lambda being the filtering function that is applied to the data
> const filteringRules = {
>   [Class.name]: (data) => data 
> };
> ```
> 
> Applying the filtering rules should be an optional parameter to `grpcUtils.fromError`. Since it used in 2 places, responding to unary calls, or responding in streams, then the parameter must be propagated upwards to `generatorWritable` and `generatorDuplex`.
> 
> The reviver and replacer are options sent to `JSON.stringify` and `JSON.parse`. These 2 functions are not called in the `ErrorPolykeyRemote`, but instead called in `grpcUtils.fromError` and `grpcUtils.toError`. These 2 rely on the `errors.ErrorPolykeyRemote` too. Therefore the `reviver` and `replacer` should be defined either in `src/errors.ts` so that these are coupled together with error definitions, or in `src/grpc/utils.ts` with the `fromError` and `toError` functions. Not sure right now, but the former looks more correct to me.

I've actually removed the `toJSON` method from `ErrorPolykey`, since I thought all of the serialisation should be done by the replacer? So I have all of the logic in there. Similarly I have no `fromJSON` methods - it's all done in the reviver.

I also don't know if filtering rules for every exception class is necessary, since the only thing we need to filter out is the stack (this is because all other information is stored in the `data` property, and we already ensure that sensitive information is never stored here when an exception is thrown in the first place.

I also thought about having the replacer/reviver in `src/errors.ts` before, however I think this would make reconstructing the errors more difficult (in the case of the reviver). If the reviver is in some other part of the codebase then we can simply do `import * as errors from 'src/errors'`, and then it's possible to do `if (value.type in errors) const error = new errors[value.type]();`

**Comment by emmacasolin** [comment]
> 6. All service handling errors should also be logged out since they don't shutdown the agent process.

I think I've covered everything except this last point, since I'm not sure how it would be done. The service handlers don't have loggers so I'm not sure how this could be achieved?

**Comment by CMCDragonkai** [comment]
> > 6. All service handling errors should also be logged out since they don't shutdown the agent process.
> 
> I think I've covered everything except this last point, since I'm not sure how it would be done. The service handlers don't have loggers so I'm not sure how this could be achieved?

Why not just inject a logger into the service handlers? Like we do with all the other things like authenticator and stuff.

**Comment by CMCDragonkai** [comment]
> I've actually removed the `toJSON` method from `ErrorPolykey`, since I thought all of the serialisation should be done by the replacer? So I have all of the logic in there. Similarly I have no `fromJSON` methods - it's all done in the reviver.

The methods are somewhat different meaning from reviver and replacer. They are just a way of ensuring the classes dictate their structure. Since `ErrorPolykey` is inherited by all the exception classes in our system, the `toJSON` method for `ErrorPolykey` is therefore shared by all the classes. Each `toJSON` only needs to care about itself (note this means the instance, not the class prototype), and when `JSON.stringify` is used, it automatically calls the `toJSON` method of every object before passing it to the `replacer` function. It's a nice separation of responsibility.

> I also don't know if filtering rules for every exception class is necessary, since the only thing we need to filter out is the stack (this is because all other information is stored in the `data` property, and we already ensure that sensitive information is never stored here when an exception is thrown in the first place.

Yep that can also work, just as long as filtering logic is separated out. I imagine something that can be class specific, and then default on some lambda. You could instead use a single lambda, like we do for the validation matching rules. And do something similar our match structure where we fall back on a default filter operation if nothing matches. See `matchSync` and `match`.
 
> I also thought about having the replacer/reviver in `src/errors.ts` before, however I think this would make reconstructing the errors more difficult (in the case of the reviver). If the reviver is in some other part of the codebase then we can simply do `import * as errors from 'src/errors'`, and then it's possible to do `if (value.type in errors) const error = new errors[value.type]();`

Well if the `toJSON` is used to provide the correct POJO output for each class, then I'm not sure you need to check classes in the `replacer`. The only place where you'd need to check for classes is for the `reviver`, and yes I can imagine in that case, you would want to use `errors[X]`. Having a `toJSON` in each class is good idea and is likely orthogonal to whether a `static fromJSON` method makes sense.

**Comment by emmacasolin** [comment]
> Why not just inject a logger into the service handlers? Like we do with all the other things like authenticator and stuff.

I've actually realised that we don't even need to inject a logger, since the grpc has its own logger that we already override. So for example we can simply do something like this:
```ts
function agentUnlock({ authenticate }: { authenticate: Authenticate }) {
  return async (
    call: grpc.ServerUnaryCall<utilsPB.EmptyMessage, utilsPB.EmptyMessage>,
    callback: grpc.sendUnaryData<utilsPB.EmptyMessage>,
  ): Promise<void> => {
    const logger = getLogger(); // get the instance of the logger
    try {
      const response = new utilsPB.EmptyMessage();
      const metadata = await authenticate(call.metadata);
      call.sendMetadata(metadata);
      callback(null, response);
      throw new Error('test');
      return;
    } catch (e) {
      callback(grpcUtils.fromError(e));
      if (logger.error) { // need this for type correctness, but should always exist
        logger.error(`${agentUnlock.name}:`, e); // log out as usual
      }
      return;
    }
  };
}
```
Which logs out `ERROR:grpc:agentUnlock: Error: test`

**Comment by CMCDragonkai** [comment]
Where does `getLogger` come from? If it's from `grpc` module itself, we shouldn't use that, because it won't go through our log formatting system. We should be centralising our logs so we can control it appropriately.

**Comment by CMCDragonkai** [comment]
Remember due the new DI pattern, injecting the logger is easy.

In fact the logger is already injected (see `client/service/index.ts` and `agent/service/index.ts`). Just add do this:

```ts
function agentUnlock({ logger, authenticate }: { authenticate: Authenticate }) {
 // ...
}
```

I already prepared it a while ago for the DI container.

**Comment by emmacasolin** [comment]
That would be the case, except that we have a `grpcSetLogger` utility that forces it to use our logger instead of its default one. This is done in `CommandPolykey.ts`:
```ts
// Set the global upstream GRPC logger
grpcSetLogger(this.logger.getChild('grpc'));
```

So calling `getLogger` in any of the client service handlers will fetch that logger. A similar thing can be done for the agent service handlers.

**Comment by CMCDragonkai** [comment]
Right but that logger is intended for grpc internals. The `logger` provided by the DI container in the service was intended for service handling logging.

**Comment by CMCDragonkai** [comment]
We are using the injected logger because it was designed for the handlers to use, the grpc logger was intended for the grpc internals to use.

**Comment by emmacasolin** [comment]
I've switched over to using injected loggers now. I just need to make sure that this logger is being injected in all the right places (where the service handlers are created) and create a utility for only logging the exceptions we want (i.e validation/authentication/precondition errors).

**Comment by CMCDragonkai** [comment]
Ok I've prototyped a little example of how `toJSON` and `fromJSON` works.

```ts
class X {
  // Polymorphic this for static members
  // https://github.com/microsoft/TypeScript/issues/5863
  public static fromJSON<T extends typeof X>(
    this: T,
    json: {
      type: string;
      data: {
        message: string;
        cause: any;
      }
    }
  ): InstanceType<T> {
    // Lack of class name type literals means we have to do runtime check
    // https://github.com/microsoft/TypeScript/issues/1579
    if (json.type !== this.name) {
      throw new Error('Incorrect JSON');
    }
    return new this(
      json.data.message,
      json.data.cause
    ) as InstanceType<T>;
  }

  public message: string;
  public cause: any;

  public constructor (message: string, cause: any) {
    this.message = 'abc';
    this.cause = cause;
  }

  public toJSON (): {
    type: string,
    data: { message: string, cause: any }
  } {
    return {
      type: this.constructor.name,
      data: {
        message: this.message,
        cause: this.cause
      }
    };
  }
}
```

Here we have a class `X` that has a static `fromJSON` and instance method of `toJSON`.

```ts
const x = new X('x message', null);
console.log(x.toJSON());
console.log(X.fromJSON(x.toJSON()));
console.log(JSON.stringify(x));
```

To use it, we rely on the `toJSON` method during `JSON.stringify`. This actually means the `replacer` function is not necessary simply because the `toJSON` method is being called to convert them into the relevant object.

Furthermore, you can see that `fromJSON` and `toJSON` are duals.

Also note that `toJSON` method does not deal with the `cause` property. It just passes it along. This is because `JSON.stringify` will also end up looking for the `toJSON` method of the cause property if any.

```ts
class Y {

  public toJSON (): {
    type: string,
    data: { }
  } {
    return {
      type: this.constructor.name,
      data: { }
    };
  }

}

const x = new X('x message', new Y());
console.log(JSON.stringify(x));
// { type: 'X', data: { message: 'abc', cause: { type: 'Y', data: {} } } }
```

Therefore the only usage of `replacer` during `JSON.stringify` would be if we wanted to change the transformation depending on the key it is assigned to, or when applied to objects that don't have their custom `toJSON` method.

This is all well and good. However the complexity starts when we are talking about `JSON.parse` and the usage of `fromJSON`.

**Comment by emmacasolin** [comment]
From the research I did when I was writing these serialisation/deserialisation methods, I could find examples of combining a `toJSON` method with a replacer function (https://www.dyn-web.com/tutorials/php-js/json/tojson.php) however, on the parsing side there's only ever mention of using the reviver (https://www.dyn-web.com/tutorials/php-js/json/reviver.php). From my own understanding, this is because `JSON.stringify` is able to see the original data type, so it knows what it's working with, so it's able to work from top to bottom. This is where a `toJSON` method can be useful since it becomes possible to serialise a data structure as a whole and include additional information/relations aside from simple key-value pairs that can be useful when deserialising later.

As for `JSON.parse` all it has access to is the serialised data string. Because of this, it has to work from bottom to top - starting with the smallest pieces of data and converting these, and then putting these smaller pieces together as it works its way up. A `fromJSON` method only becomes useful once the full data structure is pieced together and we know what type it is, because up until that point we wouldn't know which `fromJSON` to use. This is why it seems to be standard to just use a reviver function (without a `fromJSON`), with the reviver also doing the final step of creating and returning the top-level data type. This is how I've done it: https://github.com/MatrixAI/js-polykey/blob/ad979c9724f5dc32594cc70a2b0696cadcb90844/src/grpc/utils/utils.ts#L211-L328

https://github.com/MatrixAI/js-polykey/blob/ad979c9724f5dc32594cc70a2b0696cadcb90844/src/ErrorPolykey.ts#L8-L59

**Comment by CMCDragonkai** [comment]
In the case of `JSON.parse`, the parsing of the JSON string produces a JSON object. If we also have `fromJSON`, it would mean that we have to combine both functions together to make a dual of how `JSON.stringify` and `toJSON` are working together. But it's not going to be as automatic.

This is where the `reviver` comes along.

The `JSON.parse` must use a `reviver`, and that `reviver` has to use the `fromJSON` methods.

But there's a problem.

1. Unlike constructors, static methods on child classes must be substitutable with the same static method on the parent class.
2. Basically parent class static methods enforces a subtype constraint on child class static methods.
3. This means if the child class has a completely different kind of constructor, the `fromJSON` static method on the parent class will not type-check (and won't work anyway since the constructor parameters don't match).
4. If we try to override the static `fromJSON` method on the child class, we hit the problem of function subtyping. The child class method must be the same or more generic in its parameter types, but the same or more specific in its return type. So while our child `fromJSON` may return a specific child instance type, it would have to have the same or more generic "open" parameters compared to the parent `fromJSON`.

However upon trying this, it doesn't work.

```ts
class Z extends X {

  public static fromJSON<T extends typeof Z>(
    this: T,
    json: {
      type: string;
      data: {
        message: string;
        cause: any;
        num: number;
      };
    }): InstanceType<T>
  {
    return new this(
      json.data.message,
      json.data.cause,
      json.data.num
    ) as InstanceType<T>;
  }

  public num: number;

  public constructor(message: string, cause: any, num: number) {
    super(message, cause);
    this.num = num;
  }

  public toJSON(): {
    type: string,
    data: {
      message: string;
      num: number,
      cause: any
    }
  } {
    return {
      type: this.constructor.name,
      data: {
        message: this.message,
        num: this.num,
        cause: this.cause
      }
    };
  }
}

const z = new Z('some message', null, 123);
console.log(Z.fromJSON(z.toJSON()));
```

This does not type check unfortunately. It may be a combination of problems stemming from the lack of static polymorphic `this`, or other TS problems. But it appears difficult to express this. Plus even if we can express this, it would limit us to ensure that `fromJSON` on child classes were always a subtype of parent class `fromJSON`.

Right now, this means as long as all of our child classes have a compatible constructor with the parent class, then we don't need to override the  parent `fromJSON` and can continue to use it. But this means all additional parameters have to be optional.

---

So while static methods seem to lack the expressivity. There is one other solution and that's where you don't use a static method `fromJSON`, but do one of 2 things:

1. Name the `fromJSON` specific to the class, like `fromJSONX`, `fromJSONY`.
2. Give each class their own `fromJSON`, but instead of different names, just define it as a normal function.
3. Use overloaded constructors instead, because constructors are not subject to subtype constraint.

Solution 2 seems better, in that it would be possible to create a catch all `fromJSON` and specific `fromJSON` whenever the constructors are different instead.

Overloaded constructors would be a nice way of associating the `fromJSON` to each class though.

**Comment by CMCDragonkai** [comment]
@emmacasolin yes, `toJSON` and `JSON.stringify` work nicely together.

And with my previous explanation of the situation, static `fromJSON` and `JSON.parse` do not work nicely together. Primarily due to `static` factory method limitations. We want something that works like constructor inheritance (in terms of defaulting logic to the parent class when child class doesn't override), but also doesn't have the subtyping constraints that static methods have. Something that's inbetween.

In an effort to make this nicely symmetrical, a `reviver` is indeed needed, the question is how and where should the `fromJSON` logic be stored such that it is part of the exception classes and can be understood as the dual of the `toJSON` method.

**Comment by CMCDragonkai** [comment]
Let me work out a prototype for solution 2, with the reviver involved to demonstrate how the reviver's logic can be separated into each class. Since I cannot use static `fromJSON`, I will need to use an alternative methods mentioned above.

**Comment by CMCDragonkai** [comment]
@emmacasolin btw, in your replacer, since the `toJSON` gets used either way, if it is an instance of `PolykeyError`, you can just return the `value` without having to deconstruct the object to reconstruct it.

Remember `AbstractError` by itself does not have a `toJSON` method. Although with this realisation, we can definitely add it in to `js-errors` subsequently.

**Comment by emmacasolin** [comment]
> @emmacasolin btw, in your replacer, since the toJSON gets used either way, if it is an instance of PolykeyError, you can just return the value without having to deconstruct the object to reconstruct it.

Yes that's what I'm doing - I'm only deconstructing it if it's an `AbstractError`. By the time the replacer is called the `toJSON` has already been called, so if it were a polykey error it would already be a JSON object at this point and wouldn't be matched as an abstract error. That check is only there for the case of an abstract error being thrown that isn't part of the polykey exception hierarchy.

**Comment by emmacasolin** [comment]
> Let me work out a prototype for solution 2, with the reviver involved to demonstrate how the reviver's logic can be separated into each class. Since I cannot use static `fromJSON`, I will need to use an alternative methods mentioned above.

My current version of the reviver has all of the logic contained inside it, but it would be simple enough to take the logic from everything inside the `if (value.type in errors)` and put that logic inside a `fromJSON` method on `ErrorPolykey` if we want to have it defined there.

```ts
function reviver(key: string, value: any): any { 
   // If the value is an error then reconstruct it 
   if ( 
     typeof value === 'object' && 
     typeof value.type === 'string' && 
     typeof value.data === 'object' 
   ) { 
     const message = value.data.message ?? ''; 
     if (value.type in errors) { 
       const error = new errors[value.type](message, { 
         timestamp: value.data.timestamp, 
         data: value.data.data, 
         cause: value.data.cause, 
       }); 
       error.exitCode = value.data.exitCode; 
       if (value.data.stack) { 
         error.stack = value.data.stack; 
       } 
       return error; 
     } else if (value.type in otherErrors) { 
       const error = new otherErrors[value.type](message); 
       if (value.data.stack) { 
         error.stack = value.data.stack; 
       } 
       return error; 
     } else { 
       const error = new errors.ErrorPolykeyUnknown('', { data: value }); 
       if (value.data.stack) { 
         error.stack = value.data.stack; 
       } 
       return error; 
     } 
   } else if (key === '') { 
     // The value is not an error 
     const error = new errors.ErrorPolykeyUnknown('', { data: value }); 
     return error; 
   } else if (key === 'timestamp') { 
     // Encode timestamps 
     const timestampParsed = Date.parse(value); 
     if (!isNaN(timestampParsed)) { 
       return new Date(timestampParsed); 
     } else { 
       return undefined; 
     } 
   } else { 
     return value; 
   } 
 } 
```

---

Having said that though, keping all of the logic in the replacer means we can just do `const error = new errors[value.type]`, but if we had `fromJSON` methods on every error class instead then we would need a LOT of switch cases here to figure out which error it is (and thus which `fromError` to use), and also there would be a lot of repetition between each `fromJSON`. I strongly recommend not having a `fromJSON` at all - a `toJSON` + replacer and reviver is much more efficient, and it's what seems to be standard anyway.

**Comment by emmacasolin** [comment]
Here's an example of them in a `test.ts` (this uses the versions currently pushed up to #366:

```ts
import * as errors from './src/errors';
import { replacer, reviver } from './src/grpc/utils/utils';

const e1 = new errors.ErrorPolykey('some error', {
  data: { int: 1, str: 'string' },
  cause: new TypeError('some type error'),
});
const e2 = new errors.ErrorACLNodeIdMissing('some other error', { cause: e1 });

const s = JSON.stringify(e2, replacer, 2);
console.log(s);
console.log(JSON.parse(s, reviver));
```

Output:
```
{
  "type": "ErrorACLNodeIdMissing",
  "data": {
    "description": "Could not find NodeId",
    "message": "some other error",
    "exitCode": 67,
    "timestamp": "2022-05-07T04:30:15.621Z",
    "data": {},
    "cause": {
      "type": "ErrorPolykey",
      "data": {
        "description": "Polykey error",
        "message": "some error",
        "exitCode": 1,
        "timestamp": "2022-05-07T04:30:15.621Z",
        "data": {
          "int": 1,
          "str": "string"
        },
        "cause": {
          "type": "TypeError",
          "data": {
            "message": "some type error",
            "stack": "TypeError: some type error\n    at Object.<anonymous> (/home/emma/Projects/js-polykey/test.ts:6:10)\n    at Module._compile (node:internal/modules/cjs/loader:1103:14)\n    at Module.m._compile (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1371:23)\n    at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)\n    at Object.require.extensions.<computed> [as .ts] (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1374:12)\n    at Module.load (node:internal/modules/cjs/loader:981:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:822:12)\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)\n    at main (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:331:12)\n    at Object.<anonymous> (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:482:3)"
          }
        },
        "stack": "ErrorPolykey: some error\n    at Object.<anonymous> (/home/emma/Projects/js-polykey/test.ts:4:12)\n    at Module._compile (node:internal/modules/cjs/loader:1103:14)\n    at Module.m._compile (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1371:23)\n    at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)\n    at Object.require.extensions.<computed> [as .ts] (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1374:12)\n    at Module.load (node:internal/modules/cjs/loader:981:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:822:12)\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)\n    at main (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:331:12)\n    at Object.<anonymous> (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:482:3)"
      }
    },
    "stack": "ErrorACLNodeIdMissing: some other error\n    at Object.<anonymous> (/home/emma/Projects/js-polykey/test.ts:8:12)\n    at Module._compile (node:internal/modules/cjs/loader:1103:14)\n    at Module.m._compile (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1371:23)\n    at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)\n    at Object.require.extensions.<computed> [as .ts] (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1374:12)\n    at Module.load (node:internal/modules/cjs/loader:981:32)\n    at Function.Module._load (node:internal/modules/cjs/loader:822:12)\n    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)\n    at main (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:331:12)\n    at Object.<anonymous> (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:482:3)"
  }
}
ErrorACLNodeIdMissing: some other error
    at Object.<anonymous> (/home/emma/Projects/js-polykey/test.ts:8:12)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    ... 7 lines matching cause stack trace ...
    at Object.<anonymous> (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:482:3) {
  timestamp: 2022-05-07T04:30:15.621Z,
  data: {},
  cause: ErrorPolykey: some error
      at Object.<anonymous> (/home/emma/Projects/js-polykey/test.ts:4:12)
      at Module._compile (node:internal/modules/cjs/loader:1103:14)
      ... 7 lines matching cause stack trace ...
      at Object.<anonymous> (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:482:3) {
    timestamp: 2022-05-07T04:30:15.621Z,
    data: { int: 1, str: 'string' },
    cause: TypeError: some type error
        at Object.<anonymous> (/home/emma/Projects/js-polykey/test.ts:6:10)
        at Module._compile (node:internal/modules/cjs/loader:1103:14)
        at Module.m._compile (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1371:23)
        at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
        at Object.require.extensions.<computed> [as .ts] (/home/emma/Projects/js-polykey/node_modules/ts-node/src/index.ts:1374:12)
        at Module.load (node:internal/modules/cjs/loader:981:32)
        at Function.Module._load (node:internal/modules/cjs/loader:822:12)
        at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
        at main (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:331:12)
        at Object.<anonymous> (/home/emma/Projects/js-polykey/node_modules/ts-node/src/bin.ts:482:3),
    exitCode: 1
  },
  exitCode: 67
}
```

**Comment by CMCDragonkai** [comment]
We cannot use static `fromJSON` method anyway as I explained above.

But if we did have a variant of that, the `new errors[value.type]` would still work. You just do `errors[value.type].fromJSON()` instead. What changes?

**Comment by CMCDragonkai** [comment]
Some progress...

```ts
type Class<T> = new (...args: any[]) => T;

class X {
  public static fromJSON<T extends Class<any>> (
    this: T,
    json: any
  ): InstanceType<T> {
    if (
      typeof json !== 'object' ||
      json.type !== this.name ||
      typeof json.data?.message !== 'string'
    ) {
      throw new Error('Incorrect JSON');
    }
    return new this(
      json.data.message,
      json.data.cause
    );
  }

  public message: string;
  public cause: any;

  public constructor (message: string, cause: any) {
    this.message = message;
    this.cause = cause;
  }

  public toJSON (): {
    type: string,
    data: { message: string, cause: any }
  } {
    return {
      type: this.constructor.name,
      data: {
        message: this.message,
        cause: this.cause
      }
    };
  }
}

class Y extends X { }

class Z extends X {
  public static fromJSON<T extends Class<any>>(
    this: T,
    json: any
  ): InstanceType<T> {
    if (
      typeof json !== 'object' ||
      json.type !== this.name ||
      typeof json.data?.message !== 'string' ||
      typeof json.data?.num !== 'number'
    ) {
      throw new Error('Incorrect JSON');
    }
    return new this(
      json.data.message,
      json.data.cause,
      json.data.num
    );
  }

  public num: number;

  public constructor(message: string, cause: any, num: number) {
    super(message, cause);
    this.num = num;
  }

  public toJSON(): {
    type: string,
    data: {
      message: string;
      cause: any
      num: number,
    }
  } {
    return {
      type: this.constructor.name,
      data: {
        message: this.message,
        cause: this.cause,
        num: this.num,
      }
    };
  }
}

const x = new X('x message', null);

const x_: X = X.fromJSON(x.toJSON());

const y = new Y('y message', null);

const y_: Y = Y.fromJSON(y.toJSON());

const z = new Z('z message', null, 123);

const z_: Z = Z.fromJSON(z.toJSON());
```

Going to integrate your replacer and reviver now.

**Comment by CMCDragonkai** [comment]
And basically it works like this:

```ts
function replacer(key, value) {
  // normal error doesn't have toJSON
  // here we do the replacement
  if (value instanceof Error) {
    return {
      type: value.constructor.name,
      data: {
        name: value.name,
        message: value.message,
        // stack: value.stack // ignore this for simplicity
      }
    }
  } else {
    return value;
  }
}

const e = new Z(
  'z message',
  new Y(
    'y message',
    new X(
      'x message',
      new SyntaxError('oh no')
    )
  ),
  123
);

const js = JSON.stringify(e, replacer);

const revivable = {
  X,
  Y,
  Z
};

function reviver(this: any, key: string, value: any): any {
  if (
    typeof value === 'object' &&
    typeof value.type === 'string' &&
    typeof value.data === 'object'
  ) {
    if (value.type in revivable) {
      return revivable[value.type].fromJSON(value)
    } else {
      return new Error('unknown');
    }
  } else {
    return value;
  }
}

const jo = JSON.parse(js, reviver);

console.log(e);
console.log(jo);

console.log(jo instanceof Z);
console.log(jo.cause instanceof Y);
```

I haven't added all the bells and whistles you have, but this should mean I can add `toJSON` and `fromJSON` to `AbstractError`.

**Comment by CMCDragonkai** [comment]
Demonstrated here: https://github.com/MatrixAI/js-errors/pull/2

**Comment by CMCDragonkai** [comment]
https://github.com/MatrixAI/js-errors/pull/2 is now merged. And `js-errors` is now `1.1.0`.

It's important that we update all usages to this in `js-workers`, `js-async-init`, `js-async-locks`, and `js-db`.

I'll be working on `js-db`, so I'll do the update on the pending SI PR.

@emmacasolin can you update the other repos to use the latest `js-errors` version.

Note that `js-errors` is now on nodev16 and using es2021, however there's no usage of any es2021 features so the code generated should be backwards compatible. If you want you can also proceed to update `js-workers`, and `js-async-init` and `js-async-locks` to nodev16 update (as a separate commit).

Also I discovered that the `.eslintrc` needs:

```
  "env": {
    "browser": true,
    "commonjs": true,
    "es2021": true,
    "node": true,
    "jest": true
  },
```

That is `es6` can be removed. And the `ecmaVersion` is not necessary and can be removed too.

```
  "parserOptions": {
    "project": "tsconfig.json",
    "sourceType": "module"
  },
```

**Comment by CMCDragonkai** [comment]
With the new `js-errors`, we can settle the question about serialisation and deserialisation, update the `reviver` and `replacer` to be similar to the one demonstrated in `js-errors` tests about JSON serialisation and deserialisation. Additionally the "unknown" error should only be done at the top-level, and PK does still wrap all of this under `ErrorPolykeyRemote`.

Furthermore error filtering still needs to be done on 2 axes:

1. When going from agent to agent or agent to client
2. When logging out the errors on the service handlers

> We will then still need to do the error filtering in PK when going from agent to agent, but that's a separate concern. And I was thinking that the stack is always filtered even when going from agent to client. I'm also wondering if error filtering is related between the server-side logging of the errors vs the filtering of error data being sent over. It's like they are separate concerns as well, and maybe it's really that we need to distinguish client errors from server errors.

**Comment by CMCDragonkai** [comment]
Then the spec above can be updated according to the behaviour.

**Comment by CMCDragonkai** [comment]
Spec is updated. It's fine to also have some duplication between `src/client` and `src/agent` since they are different services. So each "service" can have their own set of what is considered client exceptions, and also their own set of filtering rules.

**Comment by CMCDragonkai** [comment]
@emmacasolin there's 2 kinds filtering.

1. Filtering the data that goes to the client between client service and agent service (agent service is untrusted)
   - Atm only the stack is being filtered out in the agent service, no other sensitive information exists that we know of in our exceptions
   - The `fromError` utility takes in a filtering parameter, this is being used by the agent service handler
2. Filtering which exceptions are logged out at the service handlers, only exceptions that are NOT client exceptions are logged out for both client service and agent service
   - This still is TBD, needs to have the code written into `client/utils.ts` and `agent/utils.ts`

Once `2.` is done, then tick of task 10.

**Comment by CMCDragonkai** [comment]
Note that `Concurrency Conflict` can only be added into the filtering rules for `2.` once the SI is merged into PK.

**Comment by CMCDragonkai** [comment]
When you're doing `2.`, update the spec here with the exceptions being filtered out. Logging out the exceptions should be a `logger.error`.

**Comment by CMCDragonkai** [comment]
Precondition errors are "handler specific". So the filtering rule should allow each service handler to also extend the list of "client exceptions" that are not being logged out because a sort of "Missing" or "NonExistent" exception is a client error not a server error, but only for specific handler contexts.

**Comment by emmacasolin** [comment]
Client errors not to be logged out by service handlers (WIP)

Client service:

- ErrorValidation
- ErrorClientAuthMissing
- ErrorClientAuthFormat
- ErrorClientAuthDenied
- ErrorDBTransactionConflict

Additional errors specific to handlers:

- gestaltsActionsSetByIdentity
  - ErrorGestaltsGraphIdentityIdMissing
  - ErrorGestaltsGraphNodeIdMissing
- gestaltsActionsSetByNode
  - ErrorGestaltsGraphNodeIdMissing
- gestaltsActionsUnsetByIdentity
  - ErrorGestaltsGraphIdentityIdMissing
  - ErrorGestaltsGraphNodeIdMissing
- gestaltsActionsUnsetByNode
  - ErrorGestaltsGraphNodeIdMissing
- gestaltsGestaltTrustByIdentity
  - ErrorGestaltsGraphIdentityIdMissing
  - ErrorGestaltsGraphNodeIdMissing
- gestaltsGestaltTrustByNode
  - ErrorGestaltsGraphNodeIdMissing
- identitiesAuthenticate
  - ErrorProviderMissing
- identitiesClaim
  - ErrorProviderMissing
  - ErrorProviderUnauthenticated
- identitiesInfoConnectedGet
  - ErrorProviderMissing
  - ErrorProviderUnauthenticated
  - ErrorProviderUnimplemented
- identitiesInfoGet
  - ErrorProviderMissing
  - ErrorProviderUnauthenticated
- nodesClaim
  - ErrorNodeGraphNodeIdNotFound
- nodesFind
  - ErrorNodeGraphNodeIdNotFound
- nodesPing
  - ErrorNodeGraphNodeIdNotFound
- notificationsSend
  - ErrorNodeGraphNodeIdNotFound
- vaultsClone
  - ErrorNodeGraphNodeIdNotFound
  - ErrorVaultsNameConflict
- vaultsCreate
  - ErrorVaultsVaultDefined
- vaultsDelete
  - ErrorVaultsVaultUndefined
- vaultsLog
  - ErrorVaultsVaultUndefined
  - ErrorVaultReferenceInvalid
- vaultsPermissionSet
  - ErrorVaultsVaultUndefined
  - ErrorACLNodeIdMissing
  - ErrorNodeGraphNodeIdNotFound
- vaultsPermissionUnset
  - ErrorVaultsVaultUndefined
  - ErrorGestaltsGraphNodeIdMissing
- vaultsPull
  - ErrorNodeGraphNodeIdNotFound
- vaultsRename
  - ErrorVaultsVaultDefined
  - ErrorVaultsVaultUndefined
- vaultsScan
  - ErrorNodeGraphNodeIdNotFound
- vaultsSecretsDelete
  - ErrorVaultsVaultUndefined
  - ErrorSecretsSecretUndefined
- vaultsSecretsEdit
  - ErrorVaultsVaultUndefined
  - ErrorSecretsSecretUndefined
  - ErrorVaultRemoteDefined
- vaultsSecretsGet
  - ErrorVaultsVaultUndefined
  - ErrorSecretsSecretUndefined
- vaultsSecretsList
  - ErrorVaultsVaultUndefined
- vaultsSecretsMkdir
  - ErrorVaultsVaultUndefined
  - ErrorVaultsRecursive
- vaultsSecretsNew
  - ErrorVaultsVaultUndefined
  - ErrorSecretsSecretDefined
- vaultsSecretsNewDir
  - ErrorVaultsVaultUndefined
- vaultsSecretsRename
  - ErrorVaultsVaultUndefined
  - ErrorSecretsSecretUndefined
- vaultsSecretsStat
  - ErrorVaultsVaultUndefined
  - ErrorSecretsSecretUndefined
- vaultsVersion
  - ErrorVaultsVaultUndefined
  - ErrorVaultReferenceInvalid
  - ErrorVaultReferenceMissing

Agent service:

- ErrorValidation
- ErrorDBTransactionConflict

Additional errors specific to handlers:

- nodesCrossSignClaim
  - ErrorEmptyStream
  - ErrorUndefinedSinglySignedClaim
  - ErrorUndefinedSignature
  - ErrorNodesClaimType
  - ErrorUndefinedDoublySignedClaim
- notificationsSend
  - ErrorNotificationsInvalidType
  - ErrorNotificationsValidationFailed
  - ErrorNotificationsParse
  - ErrorNotificationsPermissionsNotFound
- vaultsGitInfoGet
  - ErrorVaultsVaultUndefined
  - ErrorConnetionInfoMIssing
  - ErrorVaultsPermissionDenied
- vaultsGitPackGet
  - ErrorVaultsVaultUndefined
  - ErrorConnetionInfoMIssing
  - ErrorVaultsPermissionDenied
- vaultsScan
  - ErrorConnetionInfoMIssing
  - ErrorVaultsPermissionDenied

**Comment by emmacasolin** [comment]
Would these errors (in the context of authenticating a provider or retrieving information from a provider) be client errors or server errors?

```ts
class ErrorProviderCall<T> extends ErrorIdentities<T> {
  static description = 'Invalid response received from provider';
  exitCode = sysexits.UNAVAILABLE;
}

class ErrorProviderAuthentication<T> extends ErrorIdentities<T> {
  static description = 'Could not authenticate provider';
  exitCode = sysexits.UNKNOWN;
}
```

`ErrorProviderCall` is always thrown due to invalid data being received from an identity provider, but the reason for why invalid data was returned is unspecified.

`ErrorProviderAuthentication` is thrown when attempting to authenticate a provider. It is only thrown by `Provider.authenticate()`, and it is also the only error thrown by this method. It can be thrown for a variety of reasons including invalid data returned from the identity provider, an error coming from the identity provider, or a timeout during authentication.

**Comment by CMCDragonkai** [comment]
The `ErrorProviderCalled` is a server error.

The `ErrorProviderAuthentication` is also a server error.

They are also client errors, but because they are server errors, they should be logged out.

**Comment by emmacasolin** [comment]
I'm noticing that the vaults service handlers aren't using `validateSync` to do validation, so they would be throwing `ErrorParse` rather than `ErrorValidation`. Will fix those up to use `validateSync`.

**Comment by emmacasolin** [comment]
While I was debugging some issues with the vaults tests I was able to see the new error chain - it was definitely helpful!

```
    ERROR:createService:ErrorPolykeyRemote: Vault ID must be multibase base58btc encoded strings
    ErrorPolykeyRemote: Remote error from RPC call - Vault ID must be multibase base58btc encoded strings
      command   vaultsPull
      nodeId    v1tevljtp9f6g7rlc8mgmfcka2kqp9oduu034852he3d3ubsrn6o0
      host      127.0.0.1
      port      45327
      timestamp Mon Jun 06 2022 11:45:31 GMT+1000 (Australian Eastern Standard Time)
      cause: ErrorPolykeyRemote: Remote error from RPC call - Vault ID must be multibase base58btc encoded strings
        command vaultsGitInfoGet
        nodeId  vdfnugqe23b7ce64ruo6hfbkop4k7e5br98mml107ntue04sauusg
        host    127.0.0.1
        port    52426
        timestamp       Mon Jun 06 2022 11:45:31 GMT+1000 (Australian Eastern Standard Time)
        cause: ErrorValidation: Input data failed validation - Vault ID must be multibase base58btc encoded strings
          exitCode      65
          timestamp     Mon Jun 06 2022 11:45:31 GMT+1000 (Australian Eastern Standard Time)
```

I realised though that in this case the original error is intended for the client, not the server, but since it's wrapped in an ErrorPolykeyRemote it doesn't get caught as a client error and gets logged out by the agent. The source of the error is invalid data coming from the client, but since it doesn't get picked up until it reaches a call on a remote agent maybe it could still be relevant to the server? Not sure on this one.

**Comment by CMCDragonkai** [comment]
So this is an error from client to agent 1 to agent 2.

A validation error of agent 2 to agent 1.

I argue that agent 1 should have validated the data before passing to agent 2. This means it would have been a client error.

If agent 2 validates again and says it's wrong, then agent 1 is at fault and that's a server error.

**Comment by emmacasolin** [comment]
> So this is an error from client to agent 1 to agent 2.
> 
> A validation error of agent 2 to agent 1.
> 
> I argue that agent 1 should have validated the data before passing to agent 2. This means it would have been a client error.
> 
> If agent 2 validates again and says it's wrong, then agent 1 is at fault and that's a server error.

Agent 1 can't really validate the data, since it's possible to pass in either a Vault Name or a Vault Id. Since Agent 1 doesn't know the names of Agent 2's vaults it can't know whether a random string that is passed in is a valid vault name or not. It's not until it reaches Agent 2 and it can check that the string is not the name of any of its vaults that it can check if it's a valid Vault Id instead.

**Comment by CMCDragonkai** [comment]
Ok so it's more of the case that this vault ID is missing. This is because, we first check if there is a matching vault name on agent 2, and then check if it is a base58 encoded string. If it is not, then it's really just a missing vault. The exception should be changed to state that it is simply a missing vault.

In that case, it is a client error, and therefore agent 1 should use the special filter to say that it is in fact a client error. Right now I believe it's just an array of exception classes. I imagine to do this check, you would need to change the array of exception classes to also optionally take a chain.

```ts
Array<typeof Error | Array<typeof Error>>
```

That way if it is a single error, you just check that, if there are multiple errors, you check it in a chain.

```ts
isClientError(e, [ErrorOtherClientError, [ErrorPolykeyRemote, ErrorVaultsVaultMissing]])
```

Then you would first check if it is `ErrorPolykeyRemote`, and if true, check if cause is `ErrorVaultsVaultMissing`.


Alternatively... it maybe better that the service handler on agent 1 does a `catch`, and checks if it is in fact `ErrorVaultsVaultMissing`, and then rethrows that exception as its own, thus representing an exception override.

**Comment by emmacasolin** [comment]
> Ok so it's more of the case that this vault ID is missing. This is because, we first check if there is a matching vault name on agent 2, and then check if it is a base58 encoded string. If it is not, then it's really just a missing vault. The exception should be changed to state that it is simply a missing vault.
> 
> In that case, it is a client error, and therefore agent 1 should use the special filter to say that it is in fact a client error. Right now I believe it's just an array of exception classes. I imagine to do this check, you would need to change the array of exception classes to also optionally take a chain.
> 
> ```ts
> Array<typeof Error | Array<typeof Error>>
> ```
> 
> That way if it is a single error, you just check that, if there are multiple errors, you check it in a chain.
> 
> ```ts
> isClientError(e, [ErrorOtherClientError, [ErrorPolykeyRemote, ErrorVaultsVaultMissing]])
> ```
> 
> Then you would first check if it is `ErrorPolykeyRemote`, and if true, check if cause is `ErrorVaultsVaultMissing`.
> 
> Alternatively... it maybe better that the service handler on agent 1 does a `catch`, and checks if it is in fact `ErrorVaultsVaultMissing`, and then rethrows that exception as its own, thus representing an exception override.

Rather than having to check against an array of errors, it might be better to just iterate through the chain of the error and check each cause against the set of client errors. Otherwise it seems too narrow of a check - if the underlying error is wrapped by more than one ErrorPolykeyRemote then the check would fail.

**Comment by CMCDragonkai** [comment]
It's better to be explicit than implicit in this case though.

At any case this is merged into staging now, so we should consider this to be done.

Further improvements exists in structured logging.

**Comment by CMCDragonkai** [comment]
This is done when it was merged into staging. Further work in structured logging.
